packages/extension-app/src/background/messages.ts

```diff
@@ -0,0 +1,136 @@
+// Copyright 2018-2020 @paritytech/substrate-light-ui authors & contributors
+// This software may be modified and distributed under the terms
+// of the Apache-2.0 license. See the LICENSE file for details.
+
+import { JsonRpcRequest, JsonRpcResponse } from '@polkadot/rpc-provider/types';
+import { logger } from '@polkadot/util';
+import { kusamaCc3, WasmRpcClient, westend } from '@substrate/light';
+
+import { getClient, KUSAMA_CC3_WASM, start } from './client';
+
+export type MessageType =
+  | 'ping'
+  | 'pong'
+  | 'provider.switch'
+  | 'rpc.send'
+  | 'rpc.sendSubscribe';
+
+/**
+ * The message that the content script sends to the background script
+ */
+export interface MessageRequest {
+  jsonRpc?: JsonRpcRequest | string;
+  origin: 'content';
+  type: MessageType;
+}
+
+/**
+ * The message that the background script sends to the content script
+ */
+export interface MessageResponse {
+  jsonRpc: JsonRpcResponse | boolean;
+  origin: 'background';
+  type: MessageType;
+}
+
+const l = logger('background');
+
+function rpcProxySend(
+  client: WasmRpcClient,
+  jsonRpc: JsonRpcRequest,
+  port: browser.runtime.Port
+): void {
+  client.rpcSend(JSON.stringify(jsonRpc)).then((res: string) => {
+    try {
+      port.postMessage({
+        jsonRpc: JSON.parse(res),
+        origin: 'background',
+        type: 'rpc.send',
+      });
+    } catch (error) {
+      l.error(`rpcProxySend: Error with ${res} - ${error.message}`);
+    }
+  });
+}
+
+function rpcProxySubscribe(
+  client: WasmRpcClient,
+  jsonRpc: JsonRpcRequest,
+  port: browser.runtime.Port
+): void {
+  client.rpcSubscribe(JSON.stringify(jsonRpc), (res: string) => {
+    try {
+      port.postMessage({
+        jsonRpc: { id: jsonRpc.id, ...JSON.parse(res) },
+        origin: 'background',
+        type: 'rpc.sendSubscribe',
+      });
+    } catch (error) {
+      l.error(`rpcProxySubscribe: Error with ${res} - ${error.message}`);
+    }
+  });
+}
+/**
+ * Handle a message from the content script
+ *
+ * @param payload - The message we received from the content script
+ * @param port - The port representing the content script
+ */
+export function handler(
+  payload: MessageRequest | undefined,
+  port: browser.runtime.Port
+): void {
+  if (!payload) {
+    l.warn('handler: Received empty handler, ignoring');
+    return;
+  }
+
+  const { jsonRpc, origin, type } = payload;
+
+  if (origin !== 'content') {
+    l.warn(`handler: received message with origin ${origin}, ignorning`);
+    return;
+  }
+
+  switch (type) {
+    case 'ping':
+      return port.postMessage({
+        origin: 'background',
+        type: 'pong',
+      });
+
+    case 'provider.switch': {
+      const lightClient =
+        jsonRpc === 'kusamaCc3'
+          ? kusamaCc3.fromUrl(KUSAMA_CC3_WASM)
+          : westend.fromUrl(KUSAMA_CC3_WASM);
```

Comment:
Can you explain this case? Would it be better as `'chain.switch'` than `'provider.switch'`? And how will it work if we eventually support arbitrary chain-specs?

Reply:
The reason why I chose `provider.switch` is that a wasm light client is determined by 2 inputs:
- the chain spec
- the wasm blob

(would have chosen `chain.switch` if it was only the chain spec changing)

Right now we kind of hard-code the wasm blob (`KUSAMA_CC3_WASM`), and the chain spec can only be `'kusamaCc3' | 'westend'`. Yeah we need to allow user to upload both (maybe start with only uploading chain spec), but that's for https://github.com/paritytech/substrate-light-ui/issues/766